### PR TITLE
add Half support for unary ops on CPU

### PR DESCRIPTION
**Pull request by jiayisunx**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #99002
* __->__ #98493

Add Half support for log_sigmoid and some unary ops on CPU, including sinc, acosh, asinh, atanh, digamma, trigamma, rsqrt, acos, asin, atan, ceil, cos, erf, erfc, erfinv, exp, expml, floor, log, log10, log1p, log2, i0, round, sin, sqrt, tan, tanh, trunc, lgamma.

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @soumith @voznesenskym @penguinwu @anijain2305 @EikanWang @Guobing-Chen @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @peterbell10 @desertfire

### CODE REVIEW

**Review by jgong5 — APPROVED** [review]

File: c10/util/Half-math.h
Comment (jgong5) [inline c10/util/Half-math.h:1]: I'm wondering if we can share the implementation with bfloat16 counterpart considering they share the same implementation except for data types.

Comment (jiayisunx) [reply c10/util/Half-math.h:1]: Sharing the same implementation is possible, but c10/util/Half-inl.h and c10/util/Half.cpp are implemented separately on two data types, would it be weird to just combine c10/util/Half-math.h and c10/util/BFloat16-math.h into one file?

Comment (jgong5) [reply c10/util/Half-math.h:1]: > would it be weird to just combine c10/util/Half-math.h and c10/util/BFloat16-math.h into one file?

It depends on how similar the logic looks like. Currently, the math part is almost the same. It should be ok to dedup them.

Comment (jiayisunx) [reply c10/util/Half-math.h:1]: OK, thanks.

**Review by jgong5 — APPROVED** [review]

File: c10/util/BFloat16-math.h [hunk 1]
```diff
@@ -10,95 +11,188 @@ C10_CLANG_DIAGNOSTIC_IGNORE("-Wimplicit-float-conversion")
 
 namespace std {
 
-/// Used by vec256<c10::BFloat16>::map
-inline c10::BFloat16 acos(c10::BFloat16 a) {
+template <typename T>
+struct is_reduced_floating_point
```
Comment (mingfeima) [inline c10/util/BFloat16-math.h:15]: This has been defined in `ATen/cpu/vec/vec_base.h`, you may just include `<ATen/cpu/vec/vec.h>` to use it.

File: c10/util/BFloat16-math.h [hunk 2]
```diff
@@ -10,95 +11,188 @@ C10_CLANG_DIAGNOSTIC_IGNORE("-Wimplicit-float-conversion")
...
+    : std::integral_constant<
+          bool,
+          std::is_same<T, c10::Half>::value ||
+              std::is_same<T, c10::BFloat16>::value> {};
+
+template <
+    typename T,
+    typename std::enable_if_t<is_reduced_floating_point<T>::value, int> = 0>
```
Comment (mingfeima) [inline c10/util/BFloat16-math.h:23]: nit: use `is_reduced_floating_point_v` here and the rest of code

**Review by mingfeima — CHANGES_REQUESTED** [review]
Generally LGTM.

Just not sure whether we should overload `c10/util/BFloat16-math.h` or make a copy, say `c10/util/Half-math.h` ? @jgong5 ideas ?

**Comment by jgong5** [comment]
> Just not sure whether we should overload `c10/util/BFloat16-math.h` or make a copy, say `c10/util/Half-math.h` ? @jgong5 ideas ?

Do you mean rename the header file with a more general name? The implementation of these math functions is mostly identical for bfloat16 and half so we want to share code here.

File: test/inductor/test_torchinductor_opinfo.py [hunk 3]
```diff
@@ -186,11 +186,11 @@ def process(device_type):
     "nn.functional.avg_pool2d": {i64},
     "nn.functional.adaptive_avg_pool2d": {f16},
     "nn.functional.ctc_loss": {f32, f64},
-    "nn.functional.gaussian_nll_loss": {f32, f64},
+    "nn.functional.gaussian_nll_loss": {f16, f32, f64},
     "nn.functional.local_response_norm": {i64},
     "nn.functional.one_hot": {i64},
     "nn.functional.rrelu": {f32, f64},
-    "nn.functional.triplet_margin_with_distance_loss": {f32, f64, i32, i64},
```
Comment (EikanWang) [inline test/inductor/test_torchinductor_opinfo.py:193]: Regarding this operation and `gaussian_nll_loss`, the reason why we add `fp16` is that the two operations support FP16 now.

**Comment by jiayisunx** [comment]
I found that torch.round float16 calculation are incorrect in mps backend, and I have to add 'round' into expected fail list as workarround, and I have reported a bug issue to pytorch: https://github.com/pytorch/pytorch/issues/101167

File: c10/util/BFloat16-math.h (hunk 2 above)
Comment (jiayisunx) [reply c10/util/BFloat16-math.h]: Done, thanks.

**Review by mingfeima — APPROVED** [review]

**Comment by jiayisunx** [comment]
@ngimel, can you please review this PR? thanks!

**Review by ngimel — APPROVED** [review]

**Comment by Pitenager** [comment]
Hi team,

Do you plan to publish a new release soon?

I'm developing my MSc thesis and I'm facing this same issue `RuntimeError: "clamp_scalar_cpu" not implemented for 'Half'`

If not, is there any workaround? Downgrading is not a good one because I have other dependencies that require PyTorch >= 2.0.0

Thanks in advance

**Pull request closed without merging**